firebase-perf/src/main/java/com/google/firebase/perf/session/SessionManager.java

```diff
@@ -125,6 +125,17 @@ public void updatePerfSession(ApplicationProcessState currentAppState) {
     startOrStopCollectingGauges(currentAppState);
   }
 
+  /**
+   * Initial start of gauge collection. This should be called in ContentProvider.attachInfo during
+   * cold-start, because we want to start gauge collection as early as possible. This assumes {@link
+   * PerfSession} was already initialized a moment ago by getInstance(). Unlike updatePerfSession,
+   * this does not reset the perfSession.
+   */
+  public void initializeGaugeCollection() {
```

Comment:
By this change, we would separate SessionID creation and gaugeCollection initiation separately.

Earlier, with updatePerfSession, a new SessionID will be generated and that sessionID would be associated with all the traces and network requests that go out to the devices. With this change, though there is a sessionID, traces/network requests association to the sessionID would be lost.

Reference: https://github.com/firebase/firebase-android-sdk/blob/master/firebase-perf/src/main/java/com/google/firebase/perf/session/SessionManager.java#L115

Here is where the sessionID update is propagated to the rest of the SDK and all other traces would record this sessionID. With this new change, this update would not happen and this would make traces/network requests (active at that time - most likely app start trace) not have any session associated with it.

Can you double check if that is the case?

Reply:
This change only separate SessionID creation and gaugeCollection initiation for `FirebasePerfProvider.attachInfo`. For any `clients.updateSession(perfSession)` to be called, it has to be first added to `clients` through `registerForSessionUpdates`, and finding callers of that shows only `Trace` and `NetworkRequestMetricBuilder` registers themselves. During cold start, I doubt any Trace/NetworkTrace will be created before our `FirebasePerfProvider.attachInfo`. (`AppStartTrace` is not a `Trace`, and probably no one has network request in a ContentProvider). 

I guess it is possible that a user could've set up a high priority content provider and has a Trace in that. If you think this is possible, then I can simply add that block of code that updates `clients`, and just leave out the `PerfSession.create()`

Reply:
Sounds good. thanks for clarifying Leo. Makes sense.